src/compiler/watchUtilities.ts

```diff
@@ -70,7 +81,17 @@ namespace ts {
         }
 
         function getCachedFileSystemEntriesForBaseDir(path: Path) {
-            return getCachedFileSystemEntries(getDirectoryPath(path));
+            const entries = getCachedFileSystemEntries(getDirectoryPath(path));
+            if (!entries) {
+                return entries;
+            }
+
+            // If we're looking for the base directory, we're definitely going to search the entries
+            if (!entries.sortedAndCanonicalizedFiles) {

```

Comment:
Sorry for not being clear before but entries.files and entries.directories returned by system are sorted (not in their canonical form) https://github.com/microsoft/TypeScript/blob/main/src/compiler/sys.ts#L1842 so for write's easier to just add using `insertSorted` for it might still work to see if we can delete the entry by given name or not and if we cant then actually calculating the sorted list? 
File exists can be answered by searching with binary search on `files` and then if not found, calculating sorted list and seaching.. Also worth experimenting would be how many times do we need sorted list really with that approach.. and can the fall back be crude linear canonical comparison.. 

These are just suggestions..

Reply:
Interesting.  I didn't know we guaranteed our own sort order - I thought you meant node returned in fs order, which was generally sane.

I can give it a shot, but the code will definitely be more complicated with lazy sorting.

Reply:
@sheetalkamat Can we actually rely on the list having come from `sys`?  Can't another host return directory contents in a different order?